src/RegularizedLeastSquares.jl

```diff
@@ -20,14 +20,94 @@ using InteractiveUtils
 export AbstractLinearSolver, createLinearSolver, init, deinit, solve, linearSolverList, linearSolverListReal, applicableSolverList
 
 abstract type AbstractLinearSolver end
+
+"""
+    solve(solver::AbstractLinearSolver, b; x0 = 0, f_trace = (_, _) -> nothing)
+
+Solves an inverse problem for the data vector `b` using `solver`.
+
+# Required Arguments
+  * `solver::AbstractLinearSolver`    - linear solver (e.g., `ADMM` or `FISTA`), containing forward/normal operator and regularizer
+  * `b::AbstractVector`               - data vector if `A` was supplied to the solver, back-projection of the data otherwise
+
+# Optional Keyword Arguments
+  * `x0::AbstractVector`              - initial guess for the solution; default is zero
+  * `f_trace::Function`               - function that takes the two arguments `f_trace(solver, iteration)` and, e.g., stores, prints, or plots the intermediate solutions or convergence parameters.
+
+
+# Examples
+The optimization problem
+```math
+	argmin_x ||Ax - b||_2^2 + λ ||x||_1
+```
+can be solved with the following lines of code:
+```jldoctest solveExample
+julia> using RegularizedLeastSquares
+
+julia> A = [0.831658  0.96717
+            0.383056  0.39043
+            0.820692  0.08118];
+
+julia> x = [0.5932; 0.2697];
+
+julia> b = A * x;
+
+julia> S = ADMM(A);
+
+julia> x_approx = solve(S, b)
+2-element Vector{Float64}:
+ 0.5931999999999996
+ 0.2697000000000004
+```
+Here, we use [`L1Regularization`](@ref), which is default for [`ADMM`](@ref). All regularization options can be found in [API for Regularizers](@ref).
+
+The following example solves the same problem, but stores the solution `x` of each interation in `tr`:
+```jldoctest solveExample
+julia> tr = Dict[]
+Dict[]
+
+julia> store_trace!(tr, solver, iteration) = push!(tr, Dict("iteration" => iteration, "x" => solver.x, "beta" => solver.β))
+store_trace! (generic function with 1 method)
+
+julia> x_approx = solve(S, b; f_trace=(solver, iteration) -> store_trace!(tr, solver, iteration))
+2-element Vector{Float64}:
+ 0.5931999999999996
+ 0.2697000000000004
+
+julia> tr[3]
+Dict{String, Any} with 3 entries:
+  "iteration" => 2
+  "x"         => [0.5932, 0.2697]
+  "beta"      => [1.23143, 0.927523]
+```
+
+The last example show demonstrates how to plot the solution at every 10th iteration:
+```julia
+julia> using Plots
+
+julia> function plot_trace(solver, iteration)
+         if iteration % 10 == 0
+           display(scatter(solver.x))
+         end
+       end
+plot_trace (generic function with 1 method)
+
+julia> x_approx = solve(S, b; f_trace = plot_trace);
+```
+The keyword `f_trace` allows you to pass any function that takes the arguments `solver` and `iteration` and prints, stores, or plots intermediate result.
 """
-    solve(solver::AbstractLinearSolver, b; solverInfo = nothing, kwargs...)
+function solve(solver::AbstractLinearSolver, b; x0 = 0, f_trace = (_, _) -> nothing)
```

Comment:
@tknopp, @nHackel,
I consolidated the `solve` function as much as feasible, but I have two questions:
1. should be renamed to `solve!` as it modifies `solver`?
2. what are the use cases for the `StoreInfo` interface? I thought about an easy, yet flexible interface that I sketched here (see docstring for examples). This way, one can tailor what variables to save, print, or plot during the iterations. LMK what you think and if you feel strongly about maintaining the old interface. 

-ja

Reply:
I like the idea of providing a callback. One idea we had was providing a set of different SolverInfos, but this way we are more flexible.

And we can still provide default trace options by making structs callable: `function (info::SolverInfo)(solver, iteration) ... end`.

Something that is a little bit dangerous is that when we hand the "whole" solver to the user provided callback it could accidentally mutate the solver state. I think that is acceptable though, because that would be something very intentional by a user.

To help a user write custom traces we should provide some helper functions so that the traces don't need to be written with super specific solver knowledge. Maybe something like `tracesolution` which returns the current `x` and `traceconvergence` which returns a named tuple of convergence measures. Similarly, we could provide basic trace structs, maybe one that just traces solutions, one that traces convergence, one that traces nrmse with a provided ground truth and lastly one that allows a user to list multiple traces.

I might play around with this a little bit in a `nh/trace` branch

Reply:
I think I am in favor of renaming it to `solve!`, especially because it also mutates `x` which is a data structure that is not always initialized at the start of the solver to "clean state" but instead is something provided by the user. 

If I didn't miss anything, aside from the direct solvers only Kaczmarz has a custom solve function. I think we should move the backtransformation with the Tikhonov matrix into the `if-done` branch of the Kaczmarz iterate function. Otherwise we are in a position where calling `solve(!)` and manually using the iteration interface produces different results. 

For the tracing this would be a "jump" for the second last solution and the last solution. This might actually be another argument in favor of providing helper functions like `tracesolution` because that function for the kaczmarz solver could apply the backtransformation.

Reply:
From a naming side maybe we should go with `callback` instead of `f_trace`. It's probably a more common keyword.
And then `solverX` instead of `traceX`

Reply:
Hi @nHackel,
Thanks for your thorough feedback!  I have incorporated the following changes as you suggested:
- rename `solve` to `solve!`
- rename `f_trace` to `callback`
- added a warning sentence to the docstring to warn the user not to change the solver in the callback function. I agree that the callback interface is somewhat more advanced and I would not be worried too much that users modify the solver unintentionally. 
- removed the `storeInfo` interface. I think providing some standard functions is not a bad idea (not a must though), but I presumed this would be written a little differently as you indicate above.
- move the backtransformation with the Tikhonov matrix into the if-done branch of the Kaczmarz iterate function and remove the custom `solve!` function. 

From my end, the redesign is ready to be merged. I would like to make some improvements to the documentation and you might want to add the functionality to `createlinearsovler` and add some example callback functions, but I think this is additional functionality that can be added at any time and should not hold back a merge. 

Thoughts?